File: redo/unix/control.c
// 11 june 2015
#include "uipriv_unix.h"

#define WIDGET(c) (GTK_WIDGET(uiControlHandle((c))))

static void singleWidgetCommitDestroy(uiControl *c)
{
	g_object_unref(WIDGET(c));
}

static void singleWidgetCommitSetParent(uiControl *c, uiControl *parent)
{
	GtkWidget *newParent;
	GtkWidget *oldParent;

	if (parent == NULL) {
		oldParent = WIDGET(uiControlParent(c));
		gtk_container_remove(GTK_CONTAINER(oldParent), WIDGET(c));
		return;
	}
	newParent = WIDGET(parent);;
	gtk_container_add(GTK_CONTAINER(newParent), WIDGET(c));
}

static void singleWidgetPreferredSize(uiControl *c, uiSizing *d, intmax_t *width, intmax_t *height)
{
	// use the natural height; it makes more sense
	GtkAllocation natural;

	gtk_widget_get_preferred_size(WIDGET(c), NULL, &natural);
	*width = (intmax_t) (natural.width);
	*height = (intmax_t) (natural.height);
}

static void singleWidgetResize(uiControl *c, intmax_t x, intmax_t y, intmax_t width, intmax_t height, uiSizing *d)
{
	GtkAllocation a;

	a.x = x;
	a.y = y;
	a.width = width;
	a.height = height;
	gtk_widget_set_allocation(WIDGET(c), &a);
}

static uiSizing *singleWidgetSizing(uiControl *c)
{
	return uiUnixNewSizing();
}

static void singleWidgetCommitShow(uiControl *c)
{
	gtk_widget_show(WIDGET(c));
}

static void singleWidgetCommitHide(uiControl *c)
{
	gtk_widget_hide(WIDGET(c));
}

static void singleWidgetCommitEnable(uiControl *c)
{
	gtk_widget_set_sensitive(WIDGET(c), TRUE);
}

static void singleWidgetCommitDisable(uiControl *c)
{
	gtk_widget_set_sensitive(WIDGET(c), FALSE);
}

static uintptr_t singleWidgetStartZOrder(uiControl *c)
{
	// we don't need to do anything; GTK+ does it for us
	return 0;
}

static uintptr_t singleWidgetSetZOrder(uiControl *c, uintptr_t insertAfter)
{
	// we don't need to do anything; GTK+ does it for us
	return 0;
}

static int singleWidgetHasTabStops(uiControl *c)
{
	complain("singleWidgetHasTabStops() meaningless on GTK+");
	return 0;		// keep compiler happy
}

// called after creating the control's widget
void uiUnixMakeSingleWidgetControl(uiControl *c, GtkWidget *widget)
{
	// we have to sink the widget so we can reparent it
	g_object_ref_sink(widget);

	uiControl(c)->CommitDestroy = singleWidgetCommitDestroy;
	uiControl(c)->CommitSetParent = singleWidgetCommitSetParent;
	uiControl(c)->PreferredSize = singleWidgetPreferredSize;
	uiControl(c)->Resize = singleWidgetResize;
	uiControl(c)->Sizing = singleWidgetSizing;
	uiControl(c)->CommitShow = singleWidgetCommitShow;
	uiControl(c)->CommitHide = singleWidgetCommitHide;
	uiControl(c)->CommitEnable = singleWidgetCommitEnable;
	uiControl(c)->CommitDisable = singleWidgetCommitDisable;
	uiControl(c)->StartZOrder = singleWidgetStartZOrder;
	uiControl(c)->SetZOrder = singleWidgetSetZOrder;
	uiControl(c)->HasTabStops = singleWidgetHasTabStops;
}

void queueResize(uiControl *c)
{
	gtk_widget_queue_resize(WIDGET(c));
}

File: redo/darwin/control.m
// 11 june 2015
#include "uipriv_darwin.h"

#define VIEW(c) ((NSView *) uiControlHandle((c)))

static void singleViewCommitDestroy(uiControl *c)
{
	[VIEW(c) release];
}

static void singleViewCommitSetParent(uiControl *c, uiControl *parent)
{
	if (parent == NULL) {
		[VIEW(c) removeFromSuperview];
		return;
	}
	[VIEW(parent) addSubview:VIEW(c)];
}

static void singleViewPreferredSize(uiControl *c, uiSizing *d, intmax_t *width, intmax_t *height)
{
	NSView *v;
	NSRect r;

	v = VIEW(c);
	if (![v respondsToSelector:@selector(sizeToFit)])
		complain("uiControl %p does not respond to -sizeToFit; its type needs to provide an implementation of uiControlPreferredSize()", c);
	[v sizeToFit];
	// use alignmentRect here instead of frame because we'll be resizing based on that
	r = [v alignmentRectForFrame:[v frame]];
	*width = (intmax_t) r.size.width;
	*height = (intmax_t) r.size.height;
}

static void singleViewResize(uiControl *c, intmax_t x, intmax_t y, intmax_t width, intmax_t height, uiSizing *d)
{
	NSRect r;

	r.origin.x = x;
	// mac os x coordinate system has (0,0) in the lower-left
	r.origin.y = ([[VIEW(c) superview] bounds].size.height - height) - y;
	r.size.width = width;
	r.size.height = height;
	// this is the size of the alignment rect; the frame can be bigger (see NSButton)
	r = [VIEW(c) frameForAlignmentRect:r];
	[VIEW(c) setFrame:r];
}

static uiSizing *singleViewSizing(uiControl *c)
{
	return uiDarwinNewSizing();
}

static void singleViewCommitShow(uiControl *c)
{
	[VIEW(c) setHidden:NO];
}

static void singleViewCommitHide(uiControl *c)
{
	[VIEW(c) setHidden:YES];
}

static void singleViewCommitEnable(uiControl *c)
{
	NSControl *cc;

	if ([VIEW(c) respondsToSelector:@selector(setEnabled:)]) {
		// use NSControl to avoid compiler warnings
		cc = (NSControl *) VIEW(c);
		[cc setEnabled:YES];
	}
}

static void singleViewCommitDisable(uiControl *c)
{
	NSControl *cc;

	if ([VIEW(c) respondsToSelector:@selector(setEnabled:)]) {
		// use NSControl to avoid compiler warnings
		cc = (NSControl *) VIEW(c);
		[cc setEnabled:NO];
	}
}

static uintptr_t singleViewStartZOrder(uiControl *c)
{
	// we don't need to do anything; Cocoa does it for us
	return 0;
}

static uintptr_t singleViewSetZOrder(uiControl *c, uintptr_t insertAfter)
{
	// we don't need to do anything; Cocoa does it for us
	return 0;
}

static int singleViewHasTabStops(uiControl *c)
{
	complain("singleViewHasTabStops() meaningless on OS X");
	return 0;		// keep compiler happy
}

// called after creating the control's widget
void uiDarwinMakeSingleWidgetControl(uiControl *c, NSView *view)
{
	// we have to retain the view so we can reparent it
	[view retain];

	uiControl(c)->CommitDestroy = singleViewCommitDestroy;
	uiControl(c)->CommitSetParent = singleViewCommitSetParent;
	uiControl(c)->PreferredSize = singleViewPreferredSize;
	uiControl(c)->Resize = singleViewResize;
	uiControl(c)->Sizing = singleViewSizing;
	uiControl(c)->CommitShow = singleViewCommitShow;
	uiControl(c)->CommitHide = singleViewCommitHide;
	uiControl(c)->CommitEnable = singleViewCommitEnable;
	uiControl(c)->CommitDisable = singleViewCommitDisable;
	uiControl(c)->StartZOrder = singleViewStartZOrder;
	uiControl(c)->SetZOrder = singleViewSetZOrder;
	uiControl(c)->HasTabStops = singleViewHasTabStops;
}

void queueResize(uiControl *c)
{
	// TODO
}
